### PR TITLE
Feature/2fa api

### LINKED ISSUE #237: I scanned QR for 2FA. What's next?
image

### LINKED ISSUE #300: API endpoint for 2FA
- activation 
- check OTP

### PR DESCRIPTION
**Pull request by maksim-litvinov**
Closes #300 
Fixes #237

### CODE REVIEW

**Comment by maksim-litvinov** [comment]
Related to https://github.com/rubykube/barong/pull/277

**Comment by maksim-litvinov** [comment]
Need to move 2FA controller changes from that request to #277 as it's more relevant

**Comment by maksim-litvinov** [comment]
closes #300

File: app/controllers/api/v1/security.rb [hunk 1]
```diff
@@ -11,21 +11,72 @@ class Security < Grape::API
...
+        desc 'Verify 2FA code'
+        params do
+          requires :code, type: String, desc: 'Code from Google Authenticator',
+                          allow_blank: false
+        end
+        post '/verify_code' do
+          with_vault_error_handler do
+            error!('You need to enable 2FA first', 400) unless current_account.otp_enabled
+
+            unless Vault::TOTP.validate?(current_account.uid, params[:code])
```
Comment (andreymakovenko) [inline app/controllers/api/v1/security.rb]: use declared params

File: app/models/account.rb [hunk 2]
```diff
@@ -90,6 +90,7 @@ def random_uid
 #  state                  :string(255)      default("pending"), not null
 #  created_at             :datetime         not null
 #  updated_at             :datetime         not null
+#  otp_enabled            :boolean          default(FALSE)
```
Comment (andreymakovenko) [inline app/models/account.rb]: don't place the column at the end of database table, last 2 should be 'created_at' and 'updated_at'

File: db/migrate/20180404153832_add_otp_enabled_to_accounts.rb [hunk 3]
```diff
@@ -0,0 +1,5 @@
+class AddOtpEnabledToAccounts < ActiveRecord::Migration[5.1]
+  def change
+    add_column :accounts, :otp_enabled, :boolean, default: false
```
Comment (andreymakovenko) [inline db/migrate/20180404153832_add_otp_enabled_to_accounts.rb]: here you can use syntax like `add_column :accounts, :otp_enabled, :boolean, after: :column_name`

File: db/schema.rb [hunk 4]
```diff
@@ -101,6 +102,7 @@
     t.integer "account_id", null: false, unsigned: true
     t.datetime "created_at", null: false
     t.datetime "updated_at", null: false
+    t.string "code", limit: 5
```
Comment (andreymakovenko) [inline db/schema.rb]: what is it used for?

File: spec/api/v1/accounts_spec.rb [hunk 5]
```diff
@@ -38,21 +38,12 @@
...
-      end
-    end
-
     context 'when email and password are absent' do
       let(:params) {}
 
       it 'renders an error' do
         expect_status_to_eq 400
-        expect_body.to eq(error: 'email is missing, password is missing')
+        expect_body.to eq(error: 'email is missing, email is empty, password is missing, password is empty')
```
Comment (andreymakovenko) [inline spec/api/v1/accounts_spec.rb]: looks weird

File: spec/spec_helper.rb [hunk 6]
```diff
@@ -8,6 +8,9 @@
 
 require 'rspec/rails'
 require 'shoulda-matchers'
+require 'webmock/rspec'
+
+WebMock.allow_net_connect!
```
Comment (andreymakovenko) [inline spec/spec_helper.rb]: what is it used for?

File: db/schema.rb (hunk 4 above)
Comment (maksim-litvinov) [reply db/schema.rb]: from another request, deleted

**Pull request merged**